### ISSUE TITLE
false error when inferring a class instantiation to a union of type variables

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#3984]
### Summary

```py
class A[T]:
    t: T

    def f[R](self):
        _: A[T | R] = A()  # Object of type `A[R@f]` is not assignable to `A[T@A | R@f]`
```

[Playground](https://play.ty.dev/41a8423c-7b8d-4e69-a57a-476085446901)

**Comment by carljm** [comment]
Thanks! This is definitely a bug. It's probably related to #2817.

**Comment by carljm** [comment]
This is fixed on main.
